project/tagging/tests.py

```diff
@@ -1,3 +1,44 @@
 from django.test import TestCase
+from .models import CustomTag
 
-# Create your tests here.
+
+class CustomTagTests(TestCase):
+    def test_slugging(self):
+        test_tags = [
+            {"name": "programming", "expected_slug": "programming"},
+            {"name": "PyCon", "expected_slug": "pycon"},
+            {"name": "local storage", "expected_slug": "local-storage"},
+            {"name": "PEN testing", "expected_slug": "pen-testing"},
+            {"name": "תִּיכְנוּת", "expected_slug": "תיכנות"},
+            {"name": " 프로그램 작성", "expected_slug": "프로그램-작성"},
+            {"name": "程式设计", "expected_slug": "程式设计"},
+            {"name": "برمجة", "expected_slug": "برمجة"},
+            {"name": "आनंद", "expected_slug": "आनद"},
+            {"name": "лягушачий", "expected_slug": "лягушачий"},
+            {"name": "教程", "expected_slug": "教程"},
+            {"name": "Inicio r\u00e1pido", "expected_slug": "inicio-r\u00e1pido"},
+            {"name": "最后", "expected_slug": "最后"},
+            {"name": " 欲求不満", "expected_slug": "欲求不満"},
+            {"name": "စမ်းသပ်ခြင်း", "expected_slug": "စမသပခင"},
+            {"name": "ฐานข้อมูล", "expected_slug": "ฐานขอมล"},
+            {"name": "основы", "expected_slug": "основы"},
+            {"name": "אַלגערידאַם", "expected_slug": "אלגערידאם"},
+            {"name": "自動化する", "expected_slug": "自動化する"},
+            {"name": "sjálfvirkan", "expected_slug": "sjálfvirkan"},
+            {"name": "پژوهش ", "expected_slug": "پژوهش"},
+            {"name": " గ్రాఫ్", "expected_slug": "గరఫ"},
+            {"name": "데이터 베이스", "expected_slug": "데이터-베이스"},
+
+            {"name": "हिंदी में जानकारी", "expected_slug": "-जनकर"},
+            {"name": "प्रयास है", "expected_slug": "रयस-ह"},
+            {"name": "stòran-dàta", "expected_slug": "sstòran-dàta"},
+            {"name": "స్వయంచాలకం", "expected_slug": "యచలక"},
```

Comment:
In these four cases, I've copied the names and slugs from the fixture: https://github.com/codebuddies/backend/blob/master/project/tagging/fixtures/tagging.json but the tests fail. @bethanyg - can you confirm what the expected slugs are in these 4 cases? Have we spotted a bug, or are the fixtures misleading us here?

Reply:
So.  Two different bugs here, in different places.  Both are concerning/irritating.  

1.  The code is indeed returning **"-जनकर"** as the slug of **"हिंदी में जानकारी"** which is entirely incorrect.  In fact, **"-जनकर"** isn't a word in Hindi.  Grr.  There appears to be some hiccough in **`slugify`** that deals with spaces in Hindi words very poorly.  Additionally, the upper-case to lower-case character mapping appears to be way off.  If I attempt to slugify **"हिंदी-में-जानकारी"** ("information in Hindi"), I get back **"हद-म-जनकर"** ("half-dead"). A similar thing seems to be happening with Telugu - **"స్వయంచాలక"** ("automated") becomes **"సవయచలక"** -- which isn't a word.  However, slugifying **"базы данных"** ("database") results in the expected **"базы-данных"**.  Ditto for **"قاعدة البيانات"** ("Database"), which slugifies to **"قاعدة-البيانات"** ("Database")  _sigh_.

2.  But irrespective of slugify and it's mangling of Indic languages, we seem to have a problem with Pytest/unittest.  Because those are accurate strings for what the code is returning, so there shouldn't be a failure in matching .. unless pytest is not reading the UTF-8 correctly.

Will have to dig more..but wanted to comment on what I've found so far.

Reply:
Ugh.  CORRECTION:  The fixture **is wrong**, and loading the fixture then screws up things down the road when you try and add variants...because _uniqueness_. 🙄 

Still doesn't change the fact that the slugify code is mangling Brahmic scripts .. but I've logged that already.  For the 4 tests that aren't passing (for now -- until we fix the mangling issue), we just need to change the expected result to the one returned directly by `slugify()` in the Django shell. Will push that shortly.

Will also update the fixture file with these examples.

Reply:
Ok.  Did a commit directly here for the test fixes....wasn't exactly sure how to do it to your branch.  Oof.  Apologies.  Will dump the fixture fix into ... some subsequent PR.

Reply:
Just to zoom out slightly here.. can you explain a bit more about how you're planning to use the slugs in the application?
Is it used by the front-end? If so, how?
Are you planning to query the API by slug or GUID?
I think understanding your objective would be helpful at this point.

Reply:
Taking a stab at these questions...

> Is it used by the front-end
Yep, used by the front-end as part of URL slugs. So when a user clicks on a tag, they should be able to go to something like codebuddies.org/tags/javascript and see the resources/hangouts/learnings tagged "javascript."

> Are you planning to query the API by slug or GUID?
Hmm... I think by slug instead of by the GUID. This will happen when a user is type-searching for a slug to add to a particular resource. 

fwiw I'm not anticipating that emoji or non-English slugs at the moment, so while those are good things to file, I'm not too worried about them right now.

Reply:
> Yep, used by the front-end as part of URL

Cool, so the key thing you don't want to break is URL-safety :+1:

> This will happen when a user is type-searching for a slug to add to a particular resource

Surely the human-facing representation of the tag (and the thing you search on) is the name, not the slug?

Reply:
> Surely the human-facing representation of the tag (and the thing you search on) is the name, not the slug?

Ahh you're right, it _is_ the name! I can't think of any cases where we'd query the API by slug or GUID. 🤔

Reply:
The slug here is primarily for uniquification.  If we were searching,  we'd probably be doing some sort of similarity search across all the tags we could net for the term.  For example:  javascript would include JS, Java Script, java script, js, and quite possibly ES6 or some other terms, depending on how we implemented it.  So "clicking on the tag" would most likely not have a 1-to-1 relationship with either the name or the slug...but we **do** want a methodology of making our massive pile of tags a bit less massive/repetitive.

As for the GUID ....  I don't see a user-facing need for it.  Doesn't mean I am going to delete it from the model, but I am not seeing it as part of a URL right now.  Although if we follow our `resource` pattern, going to the GUID of a tag would let you edit that specific tag...which doesn't feel like something we want.